---
title: "Dada2 Processing: Forward Reads"
"
author: "Michael Sieler"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document: default
  pdf_document: default
---



```{r}

library(dada2); packageVersion("dada2")

```

## Paths

```{r}


# Important Paths
path.code <- paste0(proj.path,"/Code")
path.data <- paste0(proj.path,"/Data")
path.input <- paste0(proj.path,"/Data/Input") 
path.results <- paste0(proj.path,"/Results") 

## Path to data files/sub-folders
data.path <- path.data

## Path to Robjects (Renv, Rds, Rdata, etc.) saved here
objects.path <- paste0(data.path,  # Project path
                                  "/Robjects")  # Sub-directory

## Path to Output (Figures, Tables)
output.path <- paste0(data.path,  # Project path
                                 "/Output")  # Sub-directory

## Path to Output (Figures, Tables)
inDir <- paste0(data.path,  # Data path
                           "/Input")  # Sub-directory

fastq.dir <- file.path(data.path, "Raw/FastQs")

taxa.db.path <- "/Users/michaelsieler/Dropbox/Mac (2)/Documents/Sharpton_Lab/Bioinformatics_files/silva_nr99_v138.1_train_set.fa" # path to database to be used for assigning taxonomy. Should be a gzipped FASTA file.

```


```{r}

fastq.dir <- file.path(data.path, "Raw/FastQs") # CHANGE ME to the directory containing the fastq files after unzipping.

list.files(fastq.dir)

```


```{r}

# Forward and reverse fastq filenames have format: SAMPLENAME_R1_001.fastq and SAMPLENAME_R2_001.fastq
fnFs <- sort(list.files(fastq.dir, pattern="_R1_001.fastq", full.names = TRUE))
fnRs <- sort(list.files(fastq.dir, pattern="_R2_001.fastq", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 3)



```


```{r}
plotQualityProfile(fnFs[1:2])
plotQualityProfile(fnRs[1:2])

```
## Filter

```{r}

# Place filtered files in filtered/ subdirectory
filtFs <- file.path(fastq.dir, "filtered", paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(fastq.dir, "filtered", paste0(sample.names, "_R_filt.fastq.gz"))
names(filtFs) <- sample.names
names(filtRs) <- sample.names

```

```{r}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,160),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
```


### Foward reads only

```{r}
out <- filterAndTrim(fwd = fnFs, filt = filtFs, rev = NULL, filt.rev = NULL, truncLen=225,
              maxN=0, maxEE=2, truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE)

head(out)

```


## Error rates

```{r}

errF <- learnErrors(filtFs, multithread=TRUE)
plotErrors(errF, nominalQ=TRUE)

```

## Sample Inference

```{r}

dadaFs <- dada(filtFs, err=errF, multithread=TRUE)
dadaFs[[1]]

```

dada-class: object describing DADA2 denoising results
170 sequence variants were inferred from 13463 input unique sequences.
Key parameters: OMEGA_A = 1e-40, OMEGA_C = 1e-40, BAND_SIZE = 16


## Construct Sequence Table

```{r}

seqtab <- makeSequenceTable(dadaFs)
dim(seqtab)

# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))

```


## Remove Chimeras

```{r}

seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab)

```

## Track Reads

```{r}

getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "nonchim")
rownames(track) <- sample.names
head(track)

```

## Assign Taxonomy

```{r}

taxa <- assignTaxonomy(seqtab.nochim, taxa.db.path, multithread=TRUE)

```

## Inspect Taxonomy

```{r}

taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)

```

## Clean up

```{r}
sample.names[duplicated(sample.names) == TRUE]
length(sample.names)
row.names(seqtab.nochim) <-sample.names

```



## Build Tree



```{r}


ps0 <- phyloseq(
  otu_table(seqtab.nochim, taxa_are_rows = FALSE),
  sample_data(metadata),
  tax_table(taxa)
)
  
  ps1 <- numbered.ASVs(
    ps = ps0,
    # prefix = paste0(proj.name, "_ASV"),
    save.dir = output,
    save.file = "asv_sequences"
  )
  asv.seqs <- readRDS(file.path(output, "asv_sequences.rds"))

seqinr::write.fasta(
    sequences = as.list(asv.seqs),
    names = taxa_names(ps1),
    as.string = TRUE,
    file.out = file.path(output, "asv_sequences.fasta")
  )

my.cat("Proceeding with phylogenetic tree:")
    asv.seqs.file <- file.path(output, "asv_sequences.fasta")
    asv.withguides.file  <- file.path(output, "asv_and_guide_seqs.fasta")
    asv.tree.rooted.file <- file.path(output, "asv_NASTaligned_seqs.nwk")
    
    cmd <- paste(
      "cat",
      asv.seqs.file,
      guide.seqs.file,
      ">",
      asv.withguides.file
    )
    system(cmd)
    
    my.cat("Aligning sequences...")
    cmd <- paste0(
      "mothur \"#align.seqs( fasta=",
      asv.withguides.file,
      ", reference=",
      alignment.template.file,
      ", flip=t",
      ", keepdots=t",
      ", processors=", maxCores,
      ", outputdir=",
      output,
      "/ )\""
    )
    system(cmd)
    my.cat("\tDONE")
    
    mothur.output.file <- file.path(output, "asv_and_guide_seqs.align")
    fasttree.log.file <- file.path(output, "fasttree.log")
    fasttree.output.file <- file.path(output, "asv_and_guide_seqs.nwk")
    
    my.cat("Building phylogenetic tree...")
    cmd <- paste0(
      "export OMP_NUM_THREADS=",
      maxCores, "; '",
      fasttree.path, "' -nt -nosupport -quote -gtr -gamma -log ",
      fasttree.log.file,
      " ",
      mothur.output.file,
      " > ",
      fasttree.output.file
      
    )
    system(cmd)
    my.cat("\tDONE")
    asvs.and.guides.tree <- read_tree(fasttree.output.file)
    asvs.and.guides.tree.rooted <- phangorn::midpoint(asvs.and.guides.tree)
    
    guides <- scan(guide.seqs.file, what = "character" )
    guide.ids <- guides[stringr::str_detect(guides, ">" )]
    guide.ids <- stringr::str_remove(guide.ids, ">")
    
    asvs.tree.rooted <- ape::drop.tip(asvs.and.guides.tree.rooted, guide.ids)
    write.tree(asvs.tree.rooted, file = asv.tree.rooted.file)
    
    phy_tree(ps1) <- phy_tree(asvs.tree.rooted)
    system(paste("mv mothur*", output))
```

